streams-contrib/streams-persist-cassandra/src/main/java/org/apache/streams/cassandra/CassandraPersistWriter.java

```diff
@@ -0,0 +1,312 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class CassandraPersistWriter implements StreamsPersistWriter, Runnable, Flushable, Closeable {
+
+  public static final String STREAMS_ID = "CassandraPersistWriter";
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CassandraPersistWriter.class);
+
+  private static final long MAX_WRITE_LATENCY = 1000;
+
+  protected volatile Queue<StreamsDatum> persistQueue;
+
+  private ObjectMapper mapper = StreamsJacksonMapper.getInstance();
+  private volatile AtomicLong lastWrite = new AtomicLong(System.currentTimeMillis());
+  private ScheduledExecutorService backgroundFlushTask = Executors.newSingleThreadScheduledExecutor();
+
+  private CassandraConfiguration config;
+
+  protected Cluster cluster;
+  protected Session session;
+
+  protected String keyspace;
+  protected String table;
+  protected PreparedStatement insertStatement;
+
+  protected List<BoundStatement> insertBatch = new ArrayList<>();
+
+  protected final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+  public CassandraPersistWriter() {
+    this(new ComponentConfigurator<>(CassandraConfiguration.class)
+    .detectConfiguration(StreamsConfigurator.getConfig().getConfig("cassandra")));
+  }
+
+  public CassandraPersistWriter(CassandraConfiguration config) {
+    this.config = config;
+  }
+
+  public void setPersistQueue(Queue<StreamsDatum> persistQueue) {
+    this.persistQueue = persistQueue;
+  }
+
+  public Queue<StreamsDatum> getPersistQueue() {
+    return persistQueue;
+  }
+
+  @Override
+  public String getId() {
+    return STREAMS_ID;
+  }
+
+  @Override
+  public void write(StreamsDatum streamsDatum) {
+
+    ObjectNode node;
+
+    if (streamsDatum.getDocument() instanceof String) {
+      try {
+        node = mapper.readValue((String) streamsDatum.getDocument(), ObjectNode.class);
+
+        byte[] value = node.toString().getBytes();
+
+        BoundStatement statement = insertStatement.bind(GuidUtils.generateGuid(node.toString()), ByteBuffer.wrap(value));
+        insertBatch.add(statement);
+      } catch (IOException ex) {
+        ex.printStackTrace();
+        LOGGER.warn("Failure adding object: {}", streamsDatum.getDocument().toString());
+        return;
+      }
+    } else {
+      try {
+        node = mapper.valueToTree(streamsDatum.getDocument());
+
+        byte[] value = node.toString().getBytes();
+
+        BoundStatement statement = insertStatement.bind(GuidUtils.generateGuid(node.toString()), ByteBuffer.wrap(value));
```

Comment:
consider using streamsDatum.getMetadata().get("id") if it exists, enabling an stream to get a write-then-overwrite behavior when the id is already known.